profiler/src/ProfilerEngine/Datadog.Profiler.Native.Linux/LinuxStackFramesCollector.cpp

```diff
@@ -270,57 +259,63 @@ char const* LinuxStackFramesCollector::ErrorCodeToString(int32_t errorCode)
     }
 }
 
-std::int32_t LinuxStackFramesCollector::CollectCallStackCurrentThread()
+std::int32_t LinuxStackFramesCollector::CollectCallStackCurrentThread(void* ctx)
 {
     try
     {
         std::int32_t resultErrorCode;
+        // Collect data for TraceContext tracking:
+        bool traceContextDataCollected = TryApplyTraceContextDataFromCurrentCollectionThreadToSnapshot();
 
+        // if we are in the signal handler, ctx won't be null, so we will use the context
+        // This will allow us to skip the syscall frame and start from the frame before the syscall.
+        unw_cursor_t cursor;
+        if (ctx != nullptr)
         {
-            // Collect data for TraceContext tracking:
-            bool traceContextDataCollected = TryApplyTraceContextDataFromCurrentCollectionThreadToSnapshot();
-
-            // Now walk the stack:
+            resultErrorCode = unw_init_local2(&cursor, reinterpret_cast<unw_context_t*>(ctx), UNW_INIT_SIGNAL_FRAME);
+        }
+        else
+        {
+            // not in signal handler. Get the context and initialize the cursor form here
+            unw_context_t context;
+            resultErrorCode = unw_getcontext(&context);
+            if (resultErrorCode != 0)
+            {
+                return E_ABORT; // unw_getcontext does not return a specific error code. Only -1
+            }
 
-            unw_context_t uc;
-            unw_getcontext(&uc);
+            resultErrorCode = unw_init_local(&cursor, &context);
+        }
```

Comment:
This is OK, but I think you can use unw_init_local2 in both paths, and only set the flag if the context comes from the signal handler

Reply:
✔️